### DISCUSSION TITLE
Best practices for long-term sessions with 2factor

### DISCUSSION THREAD

**Discussion by Bertg** [jeremyevans/rodauth#447] [General]
Hi,

Our current rodauth setup uses the `remember` plugin to allow user to have a long-term session on their devise (this is actually a default tin our case) and we also allow users to configure several 2 factor solutions.

Now, some of our users have found that each time they kill their browser (or restart their computer) that they are confronted with a 2 factor confirmation screen again, while they expect to still be signed in.

I understand that in rodauth the current session is maintained in a session cookie and this is thus also the state of where the 2FA flag is kept, and he remember cookie is kept in a time based cookie. This explains the behaviour the users are seeing.

Now, if we would want to implement their request, what would be the best way to go about this?

* Dump the `remember` plugin and store the session in a long-lived cookie?
* Somehow add the 2FA knowledge to the `remember` state?
* Something else?

**Comment by janko** [comment]
See https://github.com/jeremyevans/rodauth/discussions/435#discussioncomment-10383625 for one possible solution.

**Comment by Bertg** [comment]
@janko Yeah, I've seen that. But it does seem a bit like a hack. In that example, if an other session would add 2FA to the account, then the first session would remain signed in via the remember function; while never having been authenticated in the first place. I would consider that a security issue. You could augment that by calling `disable_remember_login` when 2FA is added to an account, but that's a hack on top of a hack. Also, in our case, we have multiple 2FA abilities, just setting it to `totp` is a best random, at worst nonsensical.

**Reply by jeremyevans** [reply]
One reason Rodauth doesn't store two factor state in the remember token is by default, is you should consider the remember token a single factor, not multiple factors.  If the remember token automatically sets that the session is 2FA authenticated, when in reality, it is only 1FA authenticated (the remember token being the only factor), then I would consider that a security issue. 

That being said, I understand that always requiring 2FA after remember token login makes things more difficult for users.  Maybe we could have another feature that supports storing 2FA state in the remember token (either building on top of the remember feature, or as an alternative to the remember feature)?  That way, you could determine what the authentication state was at the time the remember token was created.

**Reply by Bertg** [reply]
@jeremyevans An alternative I've thought of, which is specific to our use case where we have "remember" by default, is to just drop the "remember" functionality, and instead store the session in a long lived cookie.

I thin this would simplify a lot of the setup in this scenario, and I'm pretty sure there would not be a reduction in security. 

But I'd love to hear your opinion on this.

**Reply by jeremyevans** [reply]
With that approach, you run the risk that anyone with access to the machine with the session cookie would automatically be 2FA authenticated, even if the browser is closed.  Currently, the browser has to remain open for that to be true. The default for session cookies is they are only valid until the browser is closed, and if you are using the Roda sessions plugin, a default idle/maximum limit of 7/30 days is also enforced.  Whether that is a reduction in security depends on your point of view.

**Reply by Bertg** [reply]
Understood, but this would be as secure as storing the 2FA state on the remember cookie, no?

**Reply by jeremyevans** [reply]
Long lived session cookie and 2FA state in long lived remember cookie should be equivalent in security.

**Reply by janko** [reply]
I'm wondering if all it takes to store the Rodauth session in a long-lived cookie in Rails is the following:

```rb
auth_class_eval do
  def session
    rails_cookies.permanent.signed
  end
end
```

I haven't tested it, though.

EDIT: Actually, no, that will store a new cookie for each session key. Nevermind then 😅

**Reply by Bertg** [reply]
@janko Yeah, I need to figure that one out as well. We need to figure this one out as well. Either directly via roda/rodauth or via rodauth rails.

**Reply by Bertg** [reply]
@janko with rodauth-rails I think it is as simple as adding this to the config:

```ruby
config.session_store :cookie_store,
  key: "_lol_session",
  secure: true,
  httponly: true,
  expire_after: 14.days
```

And optionally something like this to the controller:

```ruby
before_action {
  request.session_options[:expire_after] = 14.days
}
```

We are running tests with that approach now.
